rfcs/45-stageless.md

```diff
@@ -0,0 +1,984 @@
+If we removed stages, all places to evaluate run criteria and apply commands would be lost, except the beginning and end of each schedule execution.
+If we required immutable access and `bool` output from run criteria to enable basic composition, states would break because their implementation relies on run criteria that mutate a resource.
+Likewise, if we just took away stages and `ShouldRun::*AndCheckAgain`, there could be no inner frame loops (e.g. fixed timestep).
+
+Addressing even one problem involves major breaking changes.
+Ideally, we update everything in one go, as trying to spread the breaking changes over a longer period of time brings risk of adding even more technical debt.
+
+## User-facing explanation
+
+### Scheduling terminology and overview
+
+Let's define some terms.
+
+- **system**: stateful instance of a function that can access data stored in a world
+- **system set**: logical group (subgraph) of systems (can include other system sets)
+- **condition**: function that must evaluate to `true` for a system (or systems in a set) to run
+- **dependency**: system (set) that must complete before another system (set) can run
+- **schedule** (noun): the executable form of a system set
+- **schedule** (verb): specify when and under what conditions systems run
+- **executor**: runs the systems in a schedule on a world
+- **"ready"**: when a system is no longer waiting for dependencies to complete
+
+To write a Bevy app, you have to specify when your systems run.
+By default, systems have neither strict execution order nor any conditions for execution.
+**Scheduling** is the process of supplying those properties.
+To make things more ergonomic, systems can be grouped under **system sets**, which can be ordered and conditioned in the same manner as individual systems.
+This allows you to easily refer to many systems and (indirectly) give properties to many systems.
+Furthermore, systems and system sets can be ordered together and even grouped together *within larger sets*, meaning you can layer those properties.
+
+In short, for any system or system set, you can define:
+
+- its execution order relative to other systems or sets (e.g. "this system runs before A")
+- any conditions that must be true for it to run (e.g. "this system only runs if a player has full health")
+- which set(s) it belongs to (which define properties that affect all systems underneath)
+  - if left unspecified, the system or set will be added under a default one (this is configurable)
+
+These properties are all additive.
+Adding another does not replace an existing one, and they cannot be removed.
+
+### Sample
+
+```rust
+use bevy::prelude::*;
+
+#[derive(State)]
+enum GameState {
+    Running,
+    Paused
+}
+
+#[derive(SystemLabel)]
+enum MySystemSets {
+    Update,
+    Menu,
+    SubMenu,
+}
+
+#[derive(SystemLabel)]
+enum MySystems {
+    FlushMenu,
+}
+
+fn main() {
+    App::new()
+        /* ... */
+        .add_set(

```

Comment:
My vote is similarly strongly yes.

Reply:
So with this, are you also suggesting more methods for configuring a set after it's been implicitly created?

Reply:
For me at least, yes.

Reply:
> For me at least, yes.

Same. I think im proposing something along the lines of replacing `add_set()` with `configure_set(X)`.

> If you have to specify those later anyway, what's the difference? It's the same amount of typing.

This is largely a "data model" issue from my perspective. Not having a `Bar` "set object" created for a system calling `foo.in_set(Bar)` seems very weird to me, especially in a world where sets are configurable / "logically the same as schedules" / etc. In a way, we are "lying" to users about how the data is modeled. I'm happy to "lie" in the interest of UX, provided that lie is necessary and useful. Is this lie necessary and useful? Could it cause users to accidentally do something wrong? Ex: in the current system, will trying to run an "implicit set that only exists as system dependencies" as a schedule fail? If so, I would argue that this lie "isn't worth it" because that is very confusing.
(note that I've used a word with negative moral connotation, but that doesn't mean i think what you're doing here is morally wrong in any way :smile: )

Reply:
> In the current system, will trying to run an "implicit set that only exists as system dependencies" as a schedule fail?

Can you clarify this question? Do you want `in_set(X)` to not only create a set but also not require that set to be configured? If so, how is that less likely to cause accidental behavior?

In the current API, you must add a set with `add_set(X)` and for `in_set(X)` to be valid, but it doesn't matter what order they appear in, as long as you do both. Changing `add_set(X)` into `configure_set(X)` seems like a superficial detail.

I don't understand how we're "lying" about the data model. You're expressing graph relationships. If you use an unknown label in `before` and `after` in current Bevy, we don't create an empty object with that label.

Reply:
> In the current proposal, you must add a set with add_set(X) for in_set(X) to be valid, but it doesn't matter what order they appear in, as long as you do both. In that light, changing add_set(X) into configure_set(X) seems superficial (it also loses the API symmetry with systems).

Hmm I think I misinterpreted your stance. I thought you were effectively saying "creating set objects implicitly for `in_set` calls is redundant because internally the constraints will be honored per-system and the sets don't have any additional configuration by definition of the object not existing". The "lie" in that case would have been that the set object `X` doesn't necessarily exist, despite allowing `in_set(X)` to be called (and used as an ordering constraint). That model works, but I hope you can see why I thought it was weird / concerning.

But it sounds like you're actually saying "sets _only_ exist if you manually create them, you can't call `in_set(X)` if the app doesn't ultimately create the `X` set manually, and implicitly creating the set would be redundant because sets only need to exist in cases where they are manually configured".

If you are making the latter argument, I disagree. The "within a Plugin, define a set of systems without any additional config with the intention of upstream systems ordering relative to them" pattern already exists in Bevy and is something we should support.  With that model, either (1) the plugin author needs to manually call `add_set` with empty config or (2) some upstream plugin developer (or the app developer) needs to create the set manually on behalf of the plugin. Neither seems particularly desirable.

I can also see people wanting to logically group systems in sets "preemptively" as part of their app development process, prior to adding constraints, because they expect they might need to use that group in the future (and it serves as a form of documentation / organization).

Reply:
I think you're still misinterpretating what I said.

> But it sounds like you're actually saying "sets only exist if you manually create them, you can't call in_set(X) if the app doesn't ultimately create the X set manually..."

I did say this.

>"...and implicitly creating the set would be redundant because sets only need to exist in cases where they are manually configured."

I did not say this.

My comment about redundancy was more of a "Sets must be configured, so what difference does it make if we separate insertion and configuration?"

All I meant to say is that every set needs configuration. And how our builder API has always worked is you configure things *upon* insertion. I thought you had suddenly taken issue with that for some reason when it came to sets, even though it shouldn't be strange for them to share the same API systems use.

So if you're just saying "if no such set exists, create one with an empty configuration—no parents, no dependencies, no conditions" then OK, that seems like a very modest request.

But if you're arguing *in favor* of app builder methods being order-dependent such that `in_set` creates a set that needs to be scheduled further down with a `configure_set` method, that's very questionable.

> The "within a plugin, define a system set without any additional config, with the intention of ordering [outside] systems relative to them" pattern already exists in Bevy and is something we should support.

This pattern shouldn't be lost. Maybe my explanation below will make that more evident.

> With [the RFC's] model, either:
>
> 1. The plugin author needs to manually call `add_set` with empty config.
> 2. An upstream plugin developer (or the app developer) needs to create the set manually on behalf of the plugin.
>
> Neither seems particularly desirable.

I'm not sure if you're saying that users calling `add_set` with a label from a plugin just *looks* weird, but I'll respond as if you're arguing against the idea of users being able to use the `App` builder API to manipulate plugin logic.

Users being able to control where plugin systems run is *extremely* desirable (I'd put it in the top 3 things people anticipate from this refactor). It's kind of a massive problem that nearly all plugins just dump their logic in your app schedule in fixed spots and force you to design *around* them.

In this proposal, as a plugin author, you can still just shove your logic into the user's app schedule. But now you have another option. You can export the labels of certain system sets and your users can put them where they want them to go. IMO that strikes the optimal balance if your plugin's logic is very self-contained.

Case in point. Several months ago, our community favorite `bevy_rapier` merged a [change](https://github.com/dimforge/bevy_rapier/pull/169) that let users decide where its systems are inserted. I'm pretty sure that the original issue was partly prompted by our scheduling API conversation and a couple [Discord discussions with me](https://discord.com/channels/691052431525675048/742569353878437978/970075559679893594) (I linked this in the RFC), so it's kind of like a polyfill of our proposed idea.

In our proposed API, `bevy_rapier` for example could hand you a single `Physics` label and the system set that it names would be opaque to you. Keep in mind sets can be layered/nested, so plugin can both add and configure everything *below* `Physics` and still leave it to you to decide where `Physics` as a whole goes. You'd have no real way of messing with its "internals". (I think the only knob missing would be a way for plugin authors to ban adding things to their system sets.)

So in our vision, users trying to mix, say, `bevy_ggrs` and `bevy_rapier` (comes up common) could do so just by scheduling rapier's "physics" system set inside GGRS' "rollback" system set. They won't break any invariants, the plugin authors won't need to collaborate, and Bevy won't need to release a new plugin config API (and the plugin authors won't have to invent one).

To connect back to the empty config default, whether a plugin adds its a public system set with an empty config and then you change it or you just add and configure it like a system, it's the same result. Idk if you wanted to assert that the former style is nicer than the latter, but I prefer having the parity with adding systems.

> I can also see people wanting to logically group systems in sets "preemptively" as part of their app development process, prior to adding constraints, because they expect they might need to use that group in the future (and it serves as a form of documentation / organization).

So yeah, if you're just asking us to fill in missing system sets, sure, that's fine. I have to ask if it's really too much to call `add_set` with some "`nowhere`" descriptor method, but sure. If it's the other thing, I really don't think losing order-independence and parity in the API is worth saving a few keystrokes.

I will say I was confused with how you've kinda been highlighting this "empty config" usecase like it's the most important one. It's basically a brand-new concept. As far as I'm aware, nobody has been using labels (without dependencies) just to *tag* systems, since we don't have any tools to query or visualize them.

Reply:
> So if you're just saying "if no such set exists (when we start building graphs), create one with an empty configuration—no parents, no dependencies, no conditions" then OK, that seems like a very modest request.
>
> But if you're arguing in favor of in_set creating a set that you then expect to be scheduled further down with a configure_set method, that's very questionable because it reintroduces order-dependence between methods.

What I care about is (1) sets without extra configuration don't need an explicit function call to create them and (2) all "set" access apis account for this and present a coherent api.

To expand on (2): within the context of the "user-facing collection of systems, sets, and their configurations" (`Systems`, `Schedules`, `Schedule`, etc ... whatever impl we land on), I think that any logic around "retrieving systems in a set", "running systems in a set", "configuring systems in a set", etc must maintain the integrity of the data model.

So if theres some `Systems::get_systems_in_set(Bar)` function call, and previously someone called `app.add_system(foo.in_set(Bar))`, it must return `foo`, even if the schedule hasn't been run yet. As long as the user-facing data model is coherent / consistent at all times, I'm on board.

I'm against deferring data model consistency until "graph construction / schedule construction time" if that has any impact on user facing apis. Ideally our internal data model is a direct representation of our user facing data model, but if the realities of the implementation require something different, I can be convinced to make them different. But making them "different" increases the likelihood of "conceptual model desync" and it _feels_ like this particular case would benefit from doing the system<->set correlation immediately upon insert into the app.

> I'm not sure if you're saying that users calling add_set with a label from a plugin just looks weird, but I'll respond as if you're arguing against the idea of users being able to use add_set to manipulate plugin logic.

I'm not against letting users add new ordering constraints, and I also would like to support the pattern of a plugin inserting an "unconstrained" system set and letting users decide how to order it. I'm definitely not arguing against that.

I _am_ against allowing upstream plugins to _remove_ constraints, which includes allowing users to override previous set configuration. But I don't think we're disagreeing here.

I think we're both on the same side of this argument.

> I have to ask if it's really too much to call add_set with some "nowhere" descriptor method, but sure

I personally think it _is_ too much. It is unnecessary boilerplate that doesn't add any additional clarity or information. If I was a user forced to type `app.add_set(Foo)` (or `app.configure_set(Foo)`) under these circumstances, I would feel annoyed.

If a system is added to a set, that set should exist by nature of the system existing inside of the set.

I also don't see how supporting this adds order dependence:

```rust
// Permutation 1
app
    // `a` was added to the Foo set, which previously did not exist,
    // so we create it.
    .add_system(a.in_set(Foo))
    // the Foo set (which already exists) now runs after Bar
    .configure_set(Foo.after(Bar))
    // the Bar set previously did not exist, so we create it, then add it to set Baz, which also doesnt exist, so we create it
    .configure_set(Bar.in_set(Baz))
    // `b` is added to the Bar set, which already exists
    .add_system(b.in_set(Bar))
    // Foo (which already exists), is now in Baz (which already exists)
    .configure_set(Foo.in_set(Baz))

// Permutation 2
app
    // the Foo set previous did not exist, so we create it. And it now runs after Bar
    .configure_set(Foo.after(Bar))
    // Foo (which already exists), then add it to set Baz, which also doesnt exist, so we create it
    .configure_set(Foo.in_set(Baz))
    // the Bar set previously did not exist, so we create it, then add it to Baz (which already exists)
    .configure_set(Bar.in_set(Baz))
    // `a` was added to the Foo set, which already exists 
    .add_system(a.in_set(Foo))
    // `b` is added to the Bar set, which already exists
    .add_system(b.in_set(Bar))
```

There is no permutation of these lines that changes the final data representation (or schedule).

There is the open question about how to handle `Foo.after(Bar)`. This constraint could exist without the Bar set actually existing. As long as we don't implicitly assume the `Bar.before(Foo)` constraint exists in our data model, I think that is fine. But if we do, then we should create `Bar` (and add that constraint) in this case to ensure the integrity of our data model. But yeah I think I'll argue that we shouldn't create the set in that case.

Reply:
> I also don't see how supporting this adds order dependence.
> [code example]
> There is no permutation of these lines that changes the final data representation (or schedule).

Oh, I was wrong then.

> [Sets (with no config) shouldn't need an explicit function call to be created].

Yes, I agreed with this in my previous comment (well indirectly ig, I said it would be an easy addition).

> I am against allowing upstream plugins to *remove* constraints, which includes allowing users to override previous set configuration. But I don't think we're disagreeing here.

Yeah, we specifically said config methods were strictly additive. There's no notion of overriding stuff in this RFC. The only removal we spoke of was removing nodes. And when it comes to plugins, again, you would only be able to do so on the nodes the plugin decided to expose.

Only one source is allowed to call `add_set(X)` for a given `X`. I think your new axiom improves this even, since we can rely on it when a plugin wants you to call `add_set(X)` but you don't.

> As long as the user-facing data model is coherent / consistent at all times, I'm on board.

Does having many `configure_set` calls to a single set add something besides noise? Or was that just to prove the point about order-independence? If it's the latter, my strong bikeshed preference is a single call site named `add_set`.

The fact that using `X` with `in_set(X)` will be valid even if there was no `add_set(X)` doesn't mean the `add_` part is a lie, because what you have is basically a placeholder.

I do not think plugins and users should be going halfsies on setting the properties of a single system set either.

IMO both systems and sets should match in being scheduled completely in a single call site for consistent UX, and the pair of `add_system` and `add_set` just looks better.

> I'm against deferring data model consistency until "graph construction / schedule construction time" if that has any impact on user facing APIs.

What I said was probably too vague. All I meant was we won't check a graph for errors until `App::build` (at the earliest), because we can't detect cycles, ambiguities, etc. until we have all the nodes and edges. We only want to do it once anyway.

The only errors we can check for during `add_system` or `add_set` are "obvious" things like:

- you already used this name
- this node depends on itself (e.g. `add_set(X.before(X))`)
- this set contains itself (e.g. `add_set(X.in_set(X))`)
- you called `in_set(X)` but `X` is the name of a system

Reply:
> Does having many configure_set calls to a single set add something besides noise? Or was that just to prove the point about order-independence? If it's the latter, my strong bikeshed preference is a single call site named add_set.

For this specific case (to make the point about order independence) I separated them. If that block of code was in a real code base, I would combine them into a single `configure_set` call. However, I think multiple configure set calls is important to support because it allows upstream plugins to add new constraints to previously defined sets. Within that context, I don't think we need add_set / configure_set should just be used everywhere.

> I don't think plugins and users should go halfsies on setting the properties of a single system set either.

I do. If a library author is trying to export an "unconstrained" group of system functionality (ex: doesn't exist relative to anything else in the schedule), they might still want to export functionality more complicated than can be expressed in a single set.

Ex: a "calculate bounding boxes" system set and a "calculate visible entities" system set might be defined in the same plugin. "Calculate bounding boxes" must run before "calculate visible entities". But the plugin author could still leave those sets otherwise unconstrained in the schedule (with the intent to have upstream users add them somewhere). 

>  IMO both systems and sets should match in being scheduled completely in a single call site for consistent UX, and the pair of add_system and add_set just looks better.

In a world where a set can be configured exactly once, then I agree that `add_set` is the better name. But I still think we should allow configuring sets any number of times (as argued above). In that context, add_set doesn't make sense (and would be redundant with configure_set). 

> What I said was probably too vague. All I meant was we won't check a graph for errors until App::build (at the earliest), because we can't detect cycles, ambiguities, etc. until we have all the nodes and edges. We only want to do it once anyway.
The only errors we can check for during add_system or add_set are "obvious" things like: [list of things]

Good points. Agreed! Thanks for the clarity.

Reply:
> However, I think multiple configure set calls is important to support because it allows upstream plugins to add new constraints to previously defined sets. Within that context, I don't think we need add_set / configure_set should just be used everywhere.

This is my perspective too.